File: MainApp/Services/StrategyBridge.vb
Imports System.Text.RegularExpressions
Imports System.Collections.Generic
Imports System.Linq
Imports MainApp.Models

Namespace Services
    Public Class StrategyBridge
        Public Shared Function CreateFromNaturalLanguage(nlPrompt As String) As StrategyDefinition
            If String.IsNullOrWhiteSpace(nlPrompt) Then Return Nothing

            Dim text As String = NormalizeText(nlPrompt)
            Dim parts = SplitBuySell(text)

            Dim buyConditions = BuildConditions(parts.BuyText, True)
            Dim sellConditions = BuildConditions(parts.SellText, False)

            ' If user specified supertrend trend-follow but omitted sell part,
            ' provide a default exit to avoid never-ending position.
            If sellConditions.Count = 0 AndAlso MentionsSuperTrend(parts.BuyText) Then
                sellConditions.Add(New ConditionCell("S_AUTO_1", "Price CrossDown SuperTrend", "Price", ComparisonOperator.CrossDown, "SuperTrend"))
            End If

            If buyConditions.Count = 0 AndAlso sellConditions.Count = 0 Then Return Nothing

            Dim strategyName As String = "AI_Custom_" & DateTime.Now.ToString("HHmmss")
            Dim buyGate As New LogicGate("EntryGate", LogicalOperator.AND, buyConditions)
            Dim sellGate As New LogicGate("ExitGate", LogicalOperator.OR, sellConditions)

            Dim strategy As New StrategyDefinition(
                strategyName,
                "NL strategy: " & nlPrompt,
                New List(Of LogicGate) From {buyGate},
                New List(Of LogicGate) From {sellGate},
                nlPrompt)

            strategy.RequiredDataDays = 0
            Return strategy
        End Function

        Private Shared Function NormalizeText(input As String) As String
            Dim t As String = If(input, "")
            t = t.Replace(vbCr, " ").Replace(vbLf, " ").Replace("\t", " ")
            t = Regex.Replace(t, "\s+", " ")
            Return t.Trim()
        End Function

        Private Structure PromptParts
            Public BuyText As String
            Public SellText As String
        End Structure

        Private Shared Function SplitBuySell(text As String) As PromptParts
            Dim result As New PromptParts With {.BuyText = text, .SellText = ""}
            If String.IsNullOrWhiteSpace(text) Then Return result

            Dim sellKeys = New String() {"매도", "청산", "익절", "손절", "sell", "exit"}
            Dim pos As Integer = -1
            For Each k In sellKeys
                Dim p = text.IndexOf(k, StringComparison.OrdinalIgnoreCase)
                If p >= 0 AndAlso (pos = -1 OrElse p < pos) Then pos = p
            Next

            If pos > 0 Then
                result.BuyText = text.Substring(0, pos).Trim()
                result.SellText = text.Substring(pos).Trim()
            End If

            Return result
        End Function

        Private Shared Function BuildConditions(text As String, isBuy As Boolean) As List(Of ConditionCell)
            Dim conditions As New List(Of ConditionCell)
            If String.IsNullOrWhiteSpace(text) Then Return conditions

            Dim source As String = text.ToLowerInvariant()
            Dim seq As Integer = 1
            Dim prefix As String = If(isBuy, "B", "S")

            ' 1) SuperTrend templates
            ParseSuperTrend(source, isBuy, conditions, prefix, seq)

            ' 2) Tick intensity / trade strength threshold
            ParseThreshold(source, conditions, prefix, seq,
                           New String() {"틱강도", "체결강도", "tick intensity", "tick"},
                           "TICK_RAT")

            ' 3) Program net buy threshold
            ParseThreshold(source, conditions, prefix, seq,
                           New String() {"프로그램순매수", "프로그램 매매", "program net", "program"},
                           "PROGRAM_NET")

            ' 4) RSI threshold
            ParseRsi(source, conditions, prefix, seq)

            ' 5) JMA relation/cross
            ParseJma(source, isBuy, conditions, prefix, seq)

            ' 6) Risk management from sell side
            If Not isBuy Then
                ParseRisk(source, conditions, prefix, seq)
            End If

            Return Deduplicate(conditions)
        End Function

        Private Shared Sub ParseSuperTrend(source As String,
                                           isBuy As Boolean,
                                           conditions As List(Of ConditionCell),
                                           prefix As String,
                                           ByRef seq As Integer)
            If Not MentionsSuperTrend(source) Then Return

            Dim upWords = New String() {"상향", "상승", "위", "above", "long"}
            Dim downWords = New String() {"하향", "하락", "아래", "below", "short", "이탈"}
            Dim breakoutWords = New String() {"돌파", "크로스", "cross", "crossup", "crossdown"}

            Dim hasUp = upWords.Any(Function(w) source.Contains(w))
            Dim hasDown = downWords.Any(Function(w) source.Contains(w))
            Dim hasBreakout = breakoutWords.Any(Function(w) source.Contains(w))

            If isBuy Then
                Dim op = If(hasBreakout OrElse hasUp, ComparisonOperator.CrossUp, ComparisonOperator.Greater)
                AddCondition(conditions, prefix, seq,
                             "Price vs SuperTrend", "Price", op, "SuperTrend", Nothing)
            Else
                Dim op = If(hasBreakout OrElse hasDown, ComparisonOperator.CrossDown, ComparisonOperator.Less)
                AddCondition(conditions, prefix, seq,
                             "Price vs SuperTrend Exit", "Price", op, "SuperTrend", Nothing)
            End If
        End Sub

        Private Shared Function MentionsSuperTrend(source As String) As Boolean
            If String.IsNullOrWhiteSpace(source) Then Return False
            Dim s = source.ToLowerInvariant()
            Return s.Contains("supertrend") OrElse s.Contains("슈퍼트렌드") OrElse s.Contains("super trend")
        End Function

        Private Shared Sub ParseThreshold(source As String,
                                          conditions As List(Of ConditionCell),
                                          prefix As String,
                                          ByRef seq As Integer,
                                          keywords As String(),
                                          indicator As String)
            If keywords Is Nothing OrElse keywords.Length = 0 Then Return
            If Not keywords.Any(Function(k) source.Contains(k.ToLowerInvariant())) Then Return

            Dim m = Regex.Match(source, "(-?\d+(?:\.\d+)?)")
            If Not m.Success Then Return

            Dim value As Double = 0
            If Not Double.TryParse(m.Groups(1).Value, value) Then Return

            Dim op As ComparisonOperator = ComparisonOperator.GreaterEqual
            If source.Contains("이하") OrElse source.Contains("미만") OrElse source.Contains("under") OrElse source.Contains("below") Then
                op = ComparisonOperator.LessEqual
            ElseIf source.Contains("돌파") OrElse source.Contains("cross") Then
                op = ComparisonOperator.CrossUp
            End If

            AddCondition(conditions, prefix, seq,
                         indicator & " threshold", indicator, op, Nothing, value)
        End Sub

        Private Shared Sub ParseRsi(source As String,
                                    conditions As List(Of ConditionCell),
                                    prefix As String,
                                    ByRef seq As Integer)
            If Not source.Contains("rsi") Then Return

            Dim period As Integer = 14
            Dim p = Regex.Match(source, "rsi\s*(\d{1,3})")
            If p.Success Then Integer.TryParse(p.Groups(1).Value, period)

            Dim v = Regex.Match(source, "(-?\d+(?:\.\d+)?)")
            If Not v.Success Then Return

            Dim value As Double = 0
            If Not Double.TryParse(v.Groups(1).Value, value) Then Return

            Dim op As ComparisonOperator = ComparisonOperator.GreaterEqual
            If source.Contains("이하") OrElse source.Contains("under") OrElse source.Contains("below") Then
                op = ComparisonOperator.LessEqual
            ElseIf source.Contains("돌파") OrElse source.Contains("cross") Then
                op = ComparisonOperator.CrossUp
            End If

            AddCondition(conditions, prefix, seq,
                         "RSI threshold", "RSI_" & period.ToString(), op, Nothing, value)
        End Sub

        Private Shared Sub ParseJma(source As String,
                                    isBuy As Boolean,
                                    conditions As List(Of ConditionCell),
                                    prefix As String,
                                    ByRef seq As Integer)
            If Not source.Contains("jma") Then Return

            Dim period As Integer = 14
            Dim p = Regex.Match(source, "jma\s*(\d{1,3})")
            If p.Success Then Integer.TryParse(p.Groups(1).Value, period)

            Dim hasCross = source.Contains("돌파") OrElse source.Contains("크로스") OrElse source.Contains("cross")
            Dim op As ComparisonOperator

            If isBuy Then
                op = If(hasCross, ComparisonOperator.CrossUp, ComparisonOperator.Greater)
            Else
                op = If(hasCross OrElse source.Contains("이탈") OrElse source.Contains("하락"), ComparisonOperator.CrossDown, ComparisonOperator.Less)
            End If

            AddCondition(conditions, prefix, seq,
                         "Price vs JMA", "Price", op, "JMA_" & period.ToString(), Nothing)
        End Sub

        Private Shared Sub ParseRisk(source As String,
                                     conditions As List(Of ConditionCell),
                                     prefix As String,
                                     ByRef seq As Integer)
            Dim stopMatch = Regex.Match(source, "손절\s*(-?\d+(?:\.\d+)?)\s*%")
            If stopMatch.Success Then
                Dim v As Double = 0
                If Double.TryParse(stopMatch.Groups(1).Value, v) Then
                    If v > 0 Then v = -v
                    AddCondition(conditions, prefix, seq,
                                 "StopLoss", "PROFIT_PCT", ComparisonOperator.LessEqual, Nothing, v)
                End If
            End If

            Dim takeMatch = Regex.Match(source, "(익절|목표)\s*(\d+(?:\.\d+)?)\s*%")
            If takeMatch.Success Then
                Dim v As Double = 0
                If Double.TryParse(takeMatch.Groups(2).Value, v) Then
                    AddCondition(conditions, prefix, seq,
                                 "TakeProfit", "PROFIT_PCT", ComparisonOperator.GreaterEqual, Nothing, v)
                End If
            End If
        End Sub

        Private Shared Sub AddCondition(conditions As List(Of ConditionCell),
                                        prefix As String,
                                        ByRef seq As Integer,
                                        desc As String,
                                        indicatorA As String,
                                        op As ComparisonOperator,
                                        indicatorB As String,
                                        constantValue As Double?)
            Dim id As String = prefix & seq.ToString()
            seq += 1
            conditions.Add(New ConditionCell(id, desc, indicatorA, op, indicatorB, constantValue))
        End Sub

        Private Shared Function Deduplicate(conditions As List(Of ConditionCell)) As List(Of ConditionCell)
            Dim result As New List(Of ConditionCell)
            Dim keys As New HashSet(Of String)(StringComparer.OrdinalIgnoreCase)

            For Each c In conditions
                If c Is Nothing Then Continue For
                Dim k As String = String.Join("|", {
                    If(c.IndicatorA, ""),
                    c.Operator.ToString(),
                    If(c.IndicatorB, ""),
                    If(c.ConstantValue.HasValue, c.ConstantValue.Value.ToString("0.####"), "")
                })
                If keys.Contains(k) Then Continue For
                keys.Add(k)
                result.Add(c)
            Next

            Return result
        End Function
    End Class
End Namespace

File: StrategyBridge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Common.Models;

namespace App64.Services
{
    /// <summary>
    /// 사용자님의 자유로운 자연어를 정밀한 기술적 논리로 해석하는 지능형 브릿지.
    /// 정규표현식(Regex)과 키워드 매핑을 결합하여 '그냥 말하는 대로' 전략을 설계합니다.
    /// </summary>
    public static class StrategyBridge
    {
        public static StrategyDefinition CreateFromNaturalLanguage(string nlPrompt)
        {
            if (string.IsNullOrEmpty(nlPrompt)) return null;

            // 1. 매수(진입)와 매도(청산) 섹션 분리
            string buyPart = "";
            string sellPart = "";

            var splitBuy = Regex.Split(nlPrompt, "매수|진행|진입", RegexOptions.IgnoreCase);
            if (splitBuy.Length > 1)
            {
                buyPart = splitBuy[0];
                var nextPart = splitBuy[1];
                var splitSell = Regex.Split(nextPart, "매도|청산|탈출", RegexOptions.IgnoreCase);
                if (splitSell.Length > 1) { sellPart = splitSell[0]; } // 매수 뒤에 오는 매도 조건
                else { sellPart = nextPart; } // 매수 설명 이후 나머지가 매도일 가능성
            }
            else
            {
                // 구분자가 명확하지 않으면 쉼표로 분리 시도
                var clauses = nlPrompt.Split(',', '.');
                buyPart = clauses[0];
                if (clauses.Length > 1) sellPart = string.Join(",", clauses.Skip(1));
            }

            // 2. 조건 추출 및 변환
            var buyConditions = ParseConditions(buyPart, true);
            var sellConditions = ParseConditions(sellPart, false);

            if (buyConditions.Count == 0 && sellConditions.Count == 0) return null;

            // 3. 전략 조립
            string strategyName = "AI_Custom_" + DateTime.Now.ToString("HHmmss");
            var buyGate = new LogicGate("EntryGate", LogicalOperator.AND, buyConditions);
            var sellGate = new LogicGate("ExitGate", LogicalOperator.OR, sellConditions);

            // [추가] 필요 데이터 일수 계산
            int maxDays = 0;
            var allReqs = buyConditions.Concat(sellConditions).SelectMany(c => new[] { c.IndicatorA, c.IndicatorB }).Where(s => !string.IsNullOrEmpty(s));
            foreach (var req in allReqs)
            {
                // DAILY_HIGH_COND_{days}_{pct}
                var m = Regex.Match(req, @"DAILY_HIGH_COND_(\d+)_");
                if (m.Success)
                {
                    int d = int.Parse(m.Groups[1].Value);
                    if (d > maxDays) maxDays = d;
                }
            }

            return new StrategyDefinition(
                strategyName,
                "자연어 해석 전략: " + nlPrompt,
                new List<LogicGate> { buyGate },
                new List<LogicGate> { sellGate },
                nlPrompt 
            ) { RequiredDataDays = maxDays };
        }

        private static List<ConditionCell> ParseConditions(string part, bool isBuy)
        {
            var results = new List<ConditionCell>();
            if (string.IsNullOrWhiteSpace(part)) return results;

            int condId = 1;
            string prefix = isBuy ? "B" : "S";

            // [패넌 0] N일 중 ... 고가 돌파 (복합 로직 - 일봉 기준)
            var mComplex = Regex.Match(part, @"(\d+)\s*(일|봉)\s*중\s*.*(\d+)\s*%\s*이상\s*.*고가를?\s*(돌파|이상)", RegexOptions.IgnoreCase);
            if (mComplex.Success)
            {
                int days = int.Parse(mComplex.Groups[1].Value);
                int pct = int.Parse(mComplex.Groups[3].Value);
                string indicatorName = $"DAILY_HIGH_COND_{days}_{pct}";
                results.Add(new ConditionCell($"{prefix}{condId++}", $"{days}일중 {pct}%이상 상승일 고가 돌파", "Price", ComparisonOperator.CrossUp, indicatorName));
            }

            // 패턴 1: 시가대비 X% 돌파/이상/하락
            var mOpen = Regex.Match(part, @"시가대비\s*(\d+(\.\d+)?)\s*%?\s*(상승|하락)?\s*(돌파|이상|이하|초과|미만)", RegexOptions.IgnoreCase);
            if (mOpen.Success)
            {
                double val = double.Parse(mOpen.Groups[1].Value);
                if (mOpen.Groups[3].Value == "하락") val = -val;
                string opStr = mOpen.Groups[4].Value;
                results.Add(new ConditionCell($"{prefix}{condId++}", $"시가대비 {val}% {opStr}", "CHG_OPEN_PCT", MapOperator(opStr), null, val));
            }

            // 패턴 2: 틱강도 X 이상/돌파
            var mTick = Regex.Match(part, @"(틱강도|체결강도)\s*(\w+)?\s*가?\s*(\d+(\.\d+)?)\s*(이상|돌파|초과)", RegexOptions.IgnoreCase);
            if (mTick.Success)
            {
                double val = double.Parse(mTick.Groups[3].Value);
                results.Add(new ConditionCell($"{prefix}{condId++}", $"틱강도 {val} {mTick.Groups[5].Value}", "TICK_RAT", MapOperator(mTick.Groups[5].Value), null, val));
            }

            // 패턴 3: SuperTrend 상승/하락 추세
            string lowerPart = part.ToLower();
            if (lowerPart.Contains("supertrend") || lowerPart.Contains("슈퍼트렌드"))
            {
                if (lowerPart.Contains("상승추세") || lowerPart.Contains("위") || (isBuy && lowerPart.Contains("돌파")))
                    results.Add(new ConditionCell($"{prefix}{condId++}", "SuperTrend 상승 유지", "Price", ComparisonOperator.GreaterThan, "SuperTrend"));
                else if (lowerPart.Contains("하락추세") || lowerPart.Contains("아래") || (!isBuy && lowerPart.Contains("이탈")))
                    results.Add(new ConditionCell($"{prefix}{condId++}", "SuperTrend 하락 유지", "Price", ComparisonOperator.LessThan, "SuperTrend"));
            }

            // 패턴 4: 매도 특화 (VI 직전, 손절 등)
            if (!isBuy)
            {
                if (lowerPart.Contains("vi") && (lowerPart.Contains("직전") || lowerPart.Contains("근접")))
                {
                    results.Add(new ConditionCell($"{prefix}{condId++}", "VI 상한가 근접 (99% 도달)", "Price", ComparisonOperator.GreaterThanOrEqual, "VI_UP_99"));
                }

                var mStop = Regex.Match(part, @"(-?\d+)\s*%\s*(하락|이탈|손절|시)", RegexOptions.IgnoreCase);
                if (mStop.Success)
                {
                    double val = double.Parse(mStop.Groups[1].Value);
                    if (val > 0) val = -val; 
                    results.Add(new ConditionCell($"{prefix}{condId++}", $"손절매 ({val}%)", "PROFIT_PCT", ComparisonOperator.LessThanOrEqual, null, val));
                }

                var mPctRange = Regex.Match(part, @"(\d+(\.\d+)?)\s*%?\s*(상승|하락)?\s*(하면|시)", RegexOptions.IgnoreCase);
                if (mPctRange.Success)
                {
                    double val = double.Parse(mPctRange.Groups[1].Value);
                    bool isProfitTarget = part.Contains("추가") || part.Contains("수익") || part.Contains("진입");
                    
                    if (isProfitTarget)
                    {
                        if (mPctRange.Groups[3].Value != "하락")
                             results.Add(new ConditionCell($"{prefix}{condId++}", $"목표 수익률 ({val}%)", "PROFIT_PCT", ComparisonOperator.GreaterThanOrEqual, null, val));
                    }
                    else
                    {
                        if (mPctRange.Groups[3].Value == "하락")
                             results.Add(new ConditionCell($"{prefix}{condId++}", $"시가대비 {val}% 하락 매도", "CHG_OPEN_PCT", ComparisonOperator.LessThanOrEqual, null, -val));
                        else
                             results.Add(new ConditionCell($"{prefix}{condId++}", $"시가대비 {val}% 상승 매도", "CHG_OPEN_PCT", ComparisonOperator.GreaterThanOrEqual, null, val));
                    }
                }
            }

            // 패턴 5: 이평선 돌파/이탈
            var mMa = Regex.Match(part, @"(\d+)\s*(이평|MA|이동평균선)\s*(돌파|이탈|상향)", RegexOptions.IgnoreCase);
            if (mMa.Success)
            {
                string period = mMa.Groups[1].Value;
                string maName = "MA_" + period;
                string act = mMa.Groups[3].Value;
                var op = (act == "이탈") ? ComparisonOperator.CrossDown : ComparisonOperator.CrossUp;
                results.Add(new ConditionCell($"{prefix}{condId++}", $"{period}이평 {act}", "Price", op, maName));
            }

            // 패턴 6: MACD
            var mMacd = Regex.Match(part, @"(MACD)\s*(가|이)?\s*(시그널|선)?\s*(골든|데드|상향|하향)?크로스", RegexOptions.IgnoreCase);
            if (mMacd.Success)
            {
                bool isGold = part.Contains("골든") || part.Contains("상향");
                var op = isGold ? ComparisonOperator.CrossUp : ComparisonOperator.CrossDown;
                results.Add(new ConditionCell($"{prefix}{condId++}", $"MACD {(isGold ? "골든" : "데드")}크로스", "MACD_Line", op, "MACD_Signal"));
            }

            // 패턴 7: JMA
            var mJma = Regex.Match(part, @"(JMA)\s*(\d+)?\s*(상향|하향)?\s*(돌파|이탈|반전)", RegexOptions.IgnoreCase);
            if (mJma.Success)
            {
                string period = mJma.Groups[2].Success && !string.IsNullOrEmpty(mJma.Groups[2].Value) ? mJma.Groups[2].Value : "14";
                string act = mJma.Groups[4].Value;
                if (act == "반전") {
                    var isUp = part.Contains("상승") || part.Contains("상향");
                    results.Add(new ConditionCell($"{prefix}{condId++}", $"JMA({period}) {(isUp ? "상승" : "하락")}반전", $"JMA_{period}", isUp ? ComparisonOperator.CrossUp : ComparisonOperator.CrossDown, $"JMA_{period}_Prev"));
                } else {
                    var op = (act == "이탈" || mJma.Groups[3].Value == "하향") ? ComparisonOperator.CrossDown : ComparisonOperator.CrossUp;
                    results.Add(new ConditionCell($"{prefix}{condId++}", $"Price JMA({period}) {act}", "Price", op, $"JMA_{period}"));
                }
            }

            // 패턴 8: RSI
            var mRsi = Regex.Match(part, @"(RSI)\s*(\d+)?\s*(가|이)?\s*(\d+)\s*(이상|이하|돌파|이탈)", RegexOptions.IgnoreCase);
            if (mRsi.Success)
            {
                string period = mRsi.Groups[2].Success && !string.IsNullOrEmpty(mRsi.Groups[2].Value) ? mRsi.Groups[2].Value : "14";
                double val = double.Parse(mRsi.Groups[4].Value);
                string opStr = mRsi.Groups[5].Value;
                var op = MapOperator(opStr);
                results.Add(new ConditionCell($"{prefix}{condId++}", $"RSI({period}) {val} {opStr}", $"RSI_{period}", op, null, val));
            }

            // 패턴 9: 고래 체결 / 대량 거래 (Whale Flow / THI)
            var mWhale = Regex.Match(part, @"(고래|대량)\s*(매수|매도|수급)\s*(\d+)?\s*(억|백만)?\s*(이상|유입|포착)", RegexOptions.IgnoreCase);
            if (mWhale.Success)
            {
                bool isBuyWhale = mWhale.Groups[2].Value != "매도";
                double amount = mWhale.Groups[3].Success && !string.IsNullOrEmpty(mWhale.Groups[3].Value) ? double.Parse(mWhale.Groups[3].Value) : 1;
                if (mWhale.Groups[4].Value == "억") amount *= 100000000;
                
                if (part.Contains("유입") || part.Contains("포착")) {
                    results.Add(new ConditionCell($"{prefix}{condId++}", "고래 매수세 유입 (THI)", "THI_Signal", ComparisonOperator.GreaterThanOrEqual, null, 1));
                } else {
                    string ind = isBuyWhale ? "WHALE_BUY_VOL" : "WHALE_SELL_VOL";
                    results.Add(new ConditionCell($"{prefix}{condId++}", $"고래 {(isBuyWhale ? "매수" : "매도")} {mWhale.Groups[3].Value}{mWhale.Groups[4].Value} 이상", ind, ComparisonOperator.GreaterThanOrEqual, null, amount));
                }
            }

            // 패턴 10: 프로그램 순매수
            var mProg = Regex.Match(part, @"(프로그램|외인|기관)\s*(순매수|매수)?가?\s*(\d+)\s*(만주|주|억)?\s*(이상|돌파)", RegexOptions.IgnoreCase);
            if (mProg.Success)
            {
                double amount = double.Parse(mProg.Groups[3].Value);
                results.Add(new ConditionCell($"{prefix}{condId++}", $"프로그램 순매수 {amount} 이상", "PROGRAM_NET", ComparisonOperator.GreaterThanOrEqual, null, amount));
            }

            // 패턴 11: 볼린저밴드 (BB)
            var mBb = Regex.Match(part, @"(볼린저밴드|볼밴|BB)\s*(상한선|하한선|중심선|상단|하단)\s*(돌파|이탈|터치|근접)", RegexOptions.IgnoreCase);
            if (mBb.Success)
            {
                string line = mBb.Groups[2].Value;
                string act = mBb.Groups[3].Value;
                string bbVar = (line == "상한선" || line == "상단") ? "BB_UPPER" : ((line == "하한선" || line == "하단") ? "BB_LOWER" : "BB_MID");
                var op = (act == "이탈") ? ComparisonOperator.CrossDown : ComparisonOperator.CrossUp;
                if (act == "터치" || act == "근접") op = (line == "하한선" || line == "하단") ? ComparisonOperator.LessThanOrEqual : ComparisonOperator.GreaterThanOrEqual;
                
                results.Add(new ConditionCell($"{prefix}{condId++}", $"볼린저밴드 {line} {act}", "Price", op, bbVar));
            }

            return results;
        }

        private static ComparisonOperator MapOperator(string text)
        {
            if (text.Contains("돌파") || text.Contains("상향")) return ComparisonOperator.CrossUp;
            if (text.Contains("이탈") || text.Contains("하향")) return ComparisonOperator.CrossDown;
            if (text.Contains("이상") || text.Contains("초과")) return ComparisonOperator.GreaterThanOrEqual;
            if (text.Contains("이하") || text.Contains("미만")) return ComparisonOperator.LessThanOrEqual;
            return ComparisonOperator.GreaterThan;
        }
    }
}
